File: backend/src/main/scala/services/battle/microservices/extraction/objects/extraction/BattleExtractionDefinitions.scala
package services.battle.microservices.extraction.objects.extraction

import services.battle.objects.core.{BattleVector2, DurationMillis, ElapsedMillis, HeroId, PlayerId, Radius}

final case class BattleExtractionZoneId(value: String) extends AnyVal
final case class BattleLootCacheId(value: String) extends AnyVal
final case class BattleGasStageIndex(value: Int) extends AnyVal
final case class BattleGasDamagePerSecond(value: Double) extends AnyVal
final case class BattleExtractionProgressMillis(value: Long) extends AnyVal
final case class BattleLootSearchProgressMillis(value: Long) extends AnyVal
final case class BattleLootScoreValue(value: Int) extends AnyVal

enum BattleGasPhase {
  case Waiting
  case Advancing
  case Final
}

object BattleGasPhase {
  def wireValue(value: BattleGasPhase): String =
    value match {
      case BattleGasPhase.Waiting   => "waiting"
      case BattleGasPhase.Advancing => "advancing"
      case BattleGasPhase.Final     => "final"
    }
}

final case class BattleGasStageDefinition(
  stageIndex: BattleGasStageIndex,
  startsAt: ElapsedMillis,
  duration: DurationMillis,
  fromRadius: Radius,
  toRadius: Radius,
  damagePerSecond: BattleGasDamagePerSecond
)

final case class BattleGasPlanDefinition(
  center: BattleVector2,
  stages: Vector[BattleGasStageDefinition]
)

final case class BattleGasZoneState(
  phase: BattleGasPhase,
  center: BattleVector2,
  radius: Radius,
  nextRadius: Radius,
  damagePerSecond: BattleGasDamagePerSecond,
  stageIndex: BattleGasStageIndex,
  progressMs: DurationMillis,
  startsAt: ElapsedMillis,
  endsAt: ElapsedMillis
)

final case class BattleExtractionZoneDefinition(
  zoneId: BattleExtractionZoneId,
  position: BattleVector2,
  radius: Radius,
  availableFrom: ElapsedMillis,
  channelDuration: DurationMillis
)

enum BattleExtractionInterruptReason {
  case LeftZone
  case Eliminated
}

object BattleExtractionInterruptReason {
  def wireValue(value: BattleExtractionInterruptReason): String =
    value match {
      case BattleExtractionInterruptReason.LeftZone   => "left_zone"
      case BattleExtractionInterruptReason.Eliminated => "eliminated"
    }
}

enum BattleExtractionStatus {
  case Inactive
  case Available
  case Extracting(
    playerId: PlayerId,
    heroId: HeroId,
    zoneId: BattleExtractionZoneId,
    progressMs: BattleExtractionProgressMillis
  )
  case Extracted(
    playerId: PlayerId,
    heroId: HeroId,
    zoneId: BattleExtractionZoneId,
    atElapsedMs: ElapsedMillis
  )
  case Interrupted(
    playerId: PlayerId,
    heroId: HeroId,
    zoneId: BattleExtractionZoneId,
    reason: BattleExtractionInterruptReason,
    atElapsedMs: ElapsedMillis
  )
}

object BattleExtractionStatus {
  def wireValue(value: BattleExtractionStatus): String =
    value match {
      case BattleExtractionStatus.Inactive         => "inactive"
      case BattleExtractionStatus.Available        => "available"
      case BattleExtractionStatus.Extracting(_, _, _, _) => "extracting"
      case BattleExtractionStatus.Extracted(_, _, _, _)  => "extracted"
      case BattleExtractionStatus.Interrupted(_, _, _, _, _) => "interrupted"
    }
}

final case class BattleExtractionState(
  zones: Vector[BattleExtractionZoneDefinition],
  status: BattleExtractionStatus
)

final case class BattleLootCacheDefinition(
  cacheId: BattleLootCacheId,
  position: BattleVector2,
  radius: Radius,
  searchDuration: DurationMillis,
  scoreValue: BattleLootScoreValue
)

enum BattleLootCacheStatus {
  case Available
  case Searching(
    playerId: PlayerId,
    heroId: HeroId,
    progressMs: BattleLootSearchProgressMillis
  )
  case Searched(
    playerId: PlayerId,
    heroId: HeroId,
    atElapsedMs: ElapsedMillis
  )
}

object BattleLootCacheStatus {
  def wireValue(value: BattleLootCacheStatus): String =
    value match {
      case BattleLootCacheStatus.Available          => "available"
      case BattleLootCacheStatus.Searching(_, _, _) => "searching"
      case BattleLootCacheStatus.Searched(_, _, _)  => "searched"
    }
}

final case class BattleLootCacheState(
  cacheId: BattleLootCacheId,
  position: BattleVector2,
  radius: Radius,
  searchDuration: DurationMillis,
  scoreValue: BattleLootScoreValue,
  status: BattleLootCacheStatus
)

File: frontend/src/objects/battle/microservices/extraction/objects/extraction/BattleExtractionDefinitions.ts
export interface BattleVector2 {
  x: number;
  y: number;
}

export type BattleExtractionZoneId = string;
export type BattleLootCacheId = string;
export type BattleGasStageIndex = number;
export type BattleGasDamagePerSecond = number;
export type BattleExtractionProgressMillis = number;
export type BattleLootSearchProgressMillis = number;
export type BattleLootScoreValue = number;

export type BattleGasPhase = "waiting" | "advancing" | "final";

export interface BattleGasStageDefinition {
  startsAtMs: number;
  durationMs: number;
  fromRadius: number;
  toRadius: number;
  damagePerSecond: BattleGasDamagePerSecond;
}

export interface BattleGasPlanDefinition {
  center: BattleVector2;
  stages: readonly BattleGasStageDefinition[];
}

export interface BattleGasZoneState {
  phase: BattleGasPhase;
  center: BattleVector2;
  radius: number;
  nextRadius: number;
  damagePerSecond: BattleGasDamagePerSecond;
  stageIndex: BattleGasStageIndex;
  progressMs: number;
  startsAtMs: number;
  endsAtMs: number;
}

export interface BattleExtractionZoneDefinition {
  zoneId: BattleExtractionZoneId;
  position: BattleVector2;
  radius: number;
  availableFromMs: number;
  channelDurationMs: number;
}

export type BattleExtractionInterruptReason = "left_zone" | "eliminated";

export type BattleExtractionStatus =
  | { status: "inactive" }
  | { status: "available" }
  | {
      status: "extracting";
      playerId: string;
      heroId: string;
      zoneId: BattleExtractionZoneId;
      progressMs: BattleExtractionProgressMillis;
    }
  | {
      status: "extracted";
      playerId: string;
      heroId: string;
      zoneId: BattleExtractionZoneId;
      atElapsedMs: number;
    }
  | {
      status: "interrupted";
      playerId: string;
      heroId: string;
      zoneId: BattleExtractionZoneId;
      reason: BattleExtractionInterruptReason;
      atElapsedMs: number;
    };

export interface BattleExtractionState {
  zones: BattleExtractionZoneDefinition[];
  status: BattleExtractionStatus;
}

export interface BattleLootCacheDefinition {
  cacheId: BattleLootCacheId;
  position: BattleVector2;
  radius: number;
  searchDurationMs: number;
  scoreValue: BattleLootScoreValue;
}

export type BattleLootCacheStatus =
  | { status: "available" }
  | {
      status: "searching";
      playerId: string;
      heroId: string;
      progressMs: BattleLootSearchProgressMillis;
    }
  | {
      status: "searched";
      playerId: string;
      heroId: string;
      atElapsedMs: number;
    };

export interface BattleLootCacheState extends BattleLootCacheDefinition {
  status: BattleLootCacheStatus;
}

